Sources/PureSwiftJSON/Decoding/JSONUnkeyedDecodingContainer.swift

```diff
@@ -153,6 +140,34 @@ extension JSONUnkeyedDecodingContainer {
         )
     }
 
+    /// - Note: Instead of having the `isNested` parameter, it would have been quite nice to just check whether
+    ///   `T` conforms to either `KeyedDecodingContainer` or `UnkeyedDecodingContainer`. Unfortunately, since
+    ///   `KeyedDecodingContainer` takes a generic parameter (the `Key` type), we can't just ask if `T` is one, and
+    ///   type-erasure workarounds are not appropriate to this use case due to, among other things, the inability to
+    ///   conform most of the types that would matter. We also can't use `KeyedDecodingContainerProtocol` for the
+    ///   purpose, as it isn't even an existential and conformance to it can't be checked at runtime at all.
+    ///
+    ///   However, it's worth noting that the value of `isNested` is always a compile-time constant and the compiler
+    ///   can quite neatly remove whichever branch of the `if` is not taken during optimization, making doing it this
+    ///   way _much_ more performant (for what little it matters given that it's only checked in case of an error).
+    @inline(__always)
+    private func getNextValue<T>(ofType: T.Type, isNested: Bool = false) throws -> JSONValue {
+        guard !self.isAtEnd else {
+            if isNested {
+                throw DecodingError.valueNotFound(T.self,
+                                                  .init(codingPath: self.codingPath,
+                                                        debugDescription: "Cannot get nested keyed container -- unkeyed container is at end.",
+                                                        underlyingError: nil))
```

Comment:
Thanks for making the error look the same as in `Foundation` here. Though I wonder if the error message is maybe overly specific for the function name `getNextValue`.

What I actually want to say we bake in some very specific logic here that is not visible from the outside. But I can also see the value added here. What do you think about renaming it as such?

```swift
private func getNextValue<T>(ofType: T.Type, forContainer: Bool = false) throws -> JSONValue
```

@gwynne wdyt?

Reply:
@fabianfett My thinking with using the `isNested` label was that while it's true the flag also signifies whether the value will be wrapped by a container, there are _two_ ways to create containers: The `Decoder` and the  `nested[UnkeyedContainer()` methods of containers - while both methods have largely the same result, the semantics are not precisely identical. The `getNextValue()` method is specific to the unkeyed container's nested container getters - in short, I felt `isNested` best represents the specific code paths which are decided by its value. `forContainer` is ostensibly confusable with such code as `superDecoder().unkeyedContainer()`, which has closely similar effect but definitely should not share this particular decision branch.